### PR TITLE
OLMo Wrapper Part 1: Added regression , classification , and causal lm support in OLMo Wrapper

### PR DESCRIPTION
**Pull request by priyadp1**
Added a OLMo class with regression, classification, and causal lm tasks with 3 pytest files for classification, regression, and causal lm. I use a linear layer via OlmoForSequenceClassification and load the model with float16, gradient checkpointing, and quantization for memory efficiency. 

Type of change
Please check the option that is related to your PR.

 Bug fix (non-breaking change which fixes an issue)
[X ] New feature (non-breaking change which adds functionality)
 Breaking change (fix or feature that would cause existing functionality to not work as expected)
In this case, we recommend to discuss your modification on GitHub issues before creating the PR
 Documentations (modification for documents)

- [] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [X] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [X] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [X] Run `mypy -p deepchem` and check no errors
  - [X] Run `flake8 <modified file> --count` and check no errors
  - [X] Run `python -m doctest <modified file>` and check no errors
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New unit tests pass locally with my changes
- [X] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by riya-singh28** [comment]
@Harindhar10, can you please do a first review pass on this PR?

File: deepchem/models/torch_models/olmo_class.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+            if torch.cuda.is_available() else torch.float32,
+        )
+
+        super(OlmoClass, self).__init__(model=model,
+                                        tokenizer=tokenizer,
+                                        task="causal_lm",
+                                        model_dir=model_dir,
+                                        **kwargs)
+
+    def regression(self, dataset: Dataset, n_tasks: int = 1, **kwargs):
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo_class.py]: Please refer to the implementation patterns used by other models in Deepchem. Regression and classification modes do not have separate methods

File: deepchem/models/torch_models/olmo_class.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+from transformers import AutoTokenizer, AutoModelForCausalLM
+import torch
+import torch.nn as nn
+from deepchem.models.torch_models.hf_models import HuggingFaceModel
+from deepchem.data import Dataset
+
+
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo_class.py]: Please use the implementation in the following PR as a reference, as it has already gone through multiple rounds of review and aligns with Deepchem's design patterns: https://github.com/deepforestsci/chemberta4/pull/5

File: deepchem/models/torch_models/olmo_class.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+class OlmoClass(HuggingFaceModel):
+    MODEL_NAME = "allenai/OLMo-1B-hf"
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo_class.py]: Add usage examples and docstrings

File: deepchem/models/torch_models/olmo_class.py (hunk 1 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo_class.py]: @Harindhar10 Fixed that removing the classification and regression functions and replacing them with a flag wiwth default task_type being classification

File: deepchem/models/torch_models/olmo_class.py (hunk 3 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo_class.py]: @Harindhar10 Fiixed

File: deepchem/models/torch_models/olmo_class.py (hunk 2 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo_class.py]: @Harindhar10 Did the best I can to match your file. Let me know if you approve of my approach.

**Comment by priyadp1** [comment]
@riya-singh28 You can review this draft PR

File: deepchem/models/torch_models/tests/test_olmo_classification.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+SMILES = [
+    "CCN(CCSC)C(=O)N[C@@](C)(CC)C(F)(F)F",
+    "CC1(C)CN(C(=O)Nc2cc3ccccc3nn2)C[C@@]2(CCOC2)O1",
+]
+
+
+@pytest.mark.hf
+def test_olmo_single_label_classification():
+    from deepchem.models.torch_models.olmo import Olmo
+
```
Comment (riya-singh28) [inline deepchem/models/torch_models/tests/test_olmo_classification.py:24]: @priyadp1 Add a test where you load the already pretrained model using load_from_pretrained and then finetune (classification). Add a similar test for regression. Take reference from chemberta/ molformer tests.

File: deepchem/models/torch_models/tests/test_olmo_causal.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    pass
+
+SMILES = [
+    "CCN(CCSC)C(=O)N[C@@](C)(CC)C(F)(F)F",
+    "CC1(C)CN(C(=O)Nc2cc3ccccc3nn2)C[C@@]2(CCOC2)O1",
+]
+
+
+@pytest.mark.hf
+def test_olmo_causal_lm():
```
Comment (riya-singh28) [inline deepchem/models/torch_models/tests/test_olmo_causal.py]: Add an overfit test here. Take reference from other torch models.

File: deepchem/models/torch_models/olmo.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+            else:
+                if task_type in ("classification", "mtc"):
+                    problem_type = "multi_label_classification"
+                else:
+                    problem_type = "regression"
+                model = OlmoForSequenceClassification.from_pretrained(
+                    model_path,
+                    num_labels=n_tasks,
+                    problem_type=problem_type,
+                    torch_dtype=torch.bfloat16)
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo.py]: Using bfloat16 as the default is not ideal, as some hardware (e.g NVIDIA T4) does not support it and will silently fall back to FP32.

Comment (priyadp1) [reply deepchem/models/torch_models/olmo.py]: Okay, I will change that then.

File: deepchem/models/torch_models/olmo.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+        self.n_tasks = n_tasks
+        if task_type not in ("classification", "regression", "causal_lm", "mtc",
+                             "mtr"):
+            raise ValueError(
+                f"task_type must be 'classification', 'regression', 'causal_lm', 'mtc', or 'mtr', "
+                f"got '{task_type}'")
+
+        if isinstance(model, str):
+            model_path = model
+            if task_type == "causal_lm":
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo.py:190]: Deviates from the ChemBERTa and HuggingFaceModel conventions, where the model is initialized in __init__ and load_from_pretrained() is used to load pretrained weights from Hugging Face

File: deepchem/models/torch_models/tests/test_olmo_causal.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+import pytest
```
Comment (Harindhar10) [inline deepchem/models/torch_models/tests/test_olmo_causal.py:4]: Add save and reload tests

File: deepchem/models/torch_models/tests/test_olmo_classification.py (hunk 4 above)
Comment (priyadp1) [reply deepchem/models/torch_models/tests/test_olmo_classification.py:24]: @riya-singh28 added

File: deepchem/models/torch_models/tests/test_olmo_causal.py (hunk 5 above)
Comment (priyadp1) [reply deepchem/models/torch_models/tests/test_olmo_causal.py]: @riya-singh28 added

File: deepchem/models/torch_models/olmo.py (hunk 6 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo.py]: @Harindhar10 made a dictionary for different torch_dtypes so that users can choose

File: deepchem/models/torch_models/olmo.py (hunk 7 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo.py:190]: @Harindhar10 added

File: deepchem/models/torch_models/tests/test_olmo_causal.py (hunk 8 above)
Comment (priyadp1) [reply deepchem/models/torch_models/tests/test_olmo_causal.py:4]: @Harindhar10 added

**Comment by Harindhar10** [comment]
Please add tests for multi-GPU training with the LightningTorchModel class

File: deepchem/models/torch_models/olmo.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+            raise ValueError(
+                f"task_type must be 'classification', 'regression', 'causal_lm', 'mtc', or 'mtr', "
+                f"got '{task_type}'")
+
+        tokenizer = AutoTokenizer.from_pretrained(tokenizer_path)
+        olmo_config = OlmoConfig.from_pretrained(tokenizer_path)
+
+        model: Union[OlmoForCausalLM, OlmoForSequenceClassification]
+        if task_type == "causal_lm":
+            model = OlmoForCausalLM(olmo_config)
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo.py]: Is there a way to avoid initializing random weights? It takes 4–5 minutes for a 7B model, and would be infeasible for even larger models.

File: deepchem/models/torch_models/olmo.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+                problem_type = "regression"
+            self.model = OlmoForSequenceClassification.from_pretrained(
+                model_dir,
+                num_labels=self.n_tasks,
+                problem_type=problem_type,
+                quantization_config=self._quantization_config,
+                torch_dtype=self._torch_dtype,
+                low_cpu_mem_usage=True)
+
+        self.model.gradient_checkpointing_enable()
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo.py]: Please make gradient checkpointing optional

**Comment by Harindhar10** [comment]
Would be good to have support for LoRA fine-tuning

File: deepchem/models/torch_models/olmo.py (hunk 10 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo.py]: @Harindhar10 fixed

File: deepchem/models/torch_models/olmo.py (hunk 9 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo.py]: @Harindhar10 there is a library called transformers.modeling_utils that has the import no_init_weights. The no_init_weights import disables standard PyTorch weight initialization which helps large models load faster.

**Comment by riya-singh28** [comment]
@priyadp1 You can skip all the Olmo tests that are taking a lot of time to run in CI using `@pytest.mark.slow` and add a screenshot of the collab notebook (where you have run the tests and checked).

**Comment by riya-singh28** [comment]
@rbharath @priyadp1 Everything else LGTM

**Comment by priyadp1** [comment]
Screenshot 2026-07-15 at 7 01 24 PM
Screenshot 2026-07-15 at 7 01 17 PM
Screenshot 2026-07-15 at 7 01 08 PM

Proof of pytest cases passing  @rbharath @riya-singh28 @Harindhar10

File: deepchem/models/torch_models/olmo.py [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+        self.score = nn.Linear(config.hidden_size, self.num_labels, bias=False)
+        self.score = self.score.to(next(self.parameters()).dtype)
+        self.post_init()
+
+    def forward(self,
+                input_ids=None,
+                attention_mask=None,
+                labels=None,
+                **kwargs):
+        """Forward pass for OLMo sequence classification/regression."""
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo.py:58]: A more descriptive docstring with usage example is needed

File: deepchem/models/torch_models/olmo.py [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+    ----------
+    config : OlmoConfig
+        Must have num_labels and problem_type set.
+
+    Example
+    --------
+    >>> model = OlmoForSequenceClassification.from_pretrained(
+    ...     "allenai/OLMo-7B-hf", num_labels=1,
+    ...     problem_type="multi_label_classification")
+    """
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo.py:41]: Add a sample use case and not just class instantiation

File: deepchem/models/torch_models/olmo.py [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+from transformers import AutoTokenizer, AutoModel, OlmoPreTrainedModel, OlmoForCausalLM, OlmoConfig
+from transformers.modeling_outputs import SequenceClassifierOutputWithPast
+try:
+    from transformers.modeling_utils import no_init_weights
+except ImportError:
+    from transformers.initialization import no_init_weights  # type: ignore[no-redef]
+from deepchem.models.torch_models.hf_models import HuggingFaceModel
+
+
+class OlmoForSequenceClassification(OlmoPreTrainedModel):
```
Comment (Harindhar10) [inline deepchem/models/torch_models/olmo.py:15]: The latest version of Transformers (v5.14.0) already includes this class. @rbharath Would it make sense to upgrade the dependency and reuse the existing implementation? It would also make it easier to support newer state-of-the-art LLMs going forward

File: deepchem/models/torch_models/tests/test_olmo_causal.py [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+except ModuleNotFoundError:
+    pass
+
+SMILES = [
+    "CCN(CCSC)C(=O)N[C@@](C)(CC)C(F)(F)F",
+    "CC1(C)CN(C(=O)Nc2cc3ccccc3nn2)C[C@@]2(CCOC2)O1",
+]
+
+
+def quantization_config():
```
Comment (Harindhar10) [inline deepchem/models/torch_models/tests/test_olmo_causal.py]: Could this function be moved into the `HuggingFaceModel` class and made configurable through a parameter?

File: deepchem/models/torch_models/olmo.py (hunk 12 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo.py:41]: @Harindhar10 fixed

File: deepchem/models/torch_models/olmo.py (hunk 11 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo.py:58]: @Harindhar10 fixed

File: deepchem/models/torch_models/olmo.py (hunk 13 above)
Comment (priyadp1) [reply deepchem/models/torch_models/olmo.py:15]: @Harindhar10 @riya-singh28 @rbharath See PR #5088 for upgrading transformers version

File: deepchem/models/torch_models/tests/test_olmo_causal.py (hunk 14 above)
Comment (priyadp1) [reply deepchem/models/torch_models/tests/test_olmo_causal.py]: @riya-singh28  @Harindhar10 added def quantization_config to conftest.py to avoid repetition across files

**Review by riya-singh28 — APPROVED** [review]
LGTM

**Comment by Harindhar10** [comment]
LGTM

**Review by Shriprasad-P — COMMENTED** [review]
## Review

**PR:** OLMo Wrapper Part 1: Added regression , classification , and causal lm support in OLMo Wrapper

**Touched:** `deepchem/models/__init__.py`, `deepchem/models/torch_models/__init__.py`, `deepchem/models/torch_models/olmo.py`, `deepchem/models/torch_models/tests/conftest.py`, `deepchem/models/torch_models/tests/test_olmo_causal.py`, `deepchem/models/torch_models/tests/test_olmo_classification.py`

- Addresses a concrete correctness/reliability issue based on the title and diff.
- Diff is fairly large (+1110/-1); a short summary of risk areas from the author would help reviewers.

Commenting as a drive-by reviewer after reading the diff. Happy to look again if maintainers want a deeper pass on a specific file.
